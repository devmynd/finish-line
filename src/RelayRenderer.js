import React, { Component } from 'react'
import { QueryRenderer } from 'react-relay'
import PropTypes from 'prop-types'

export class RelayRenderer extends Component {
  static propTypes = {
    container: PropTypes.any,
    error: PropTypes.any,
    loading: PropTypes.any,
    query: PropTypes.any.isRequired,
    refreshRelayEnvironment: PropTypes.any,
    render: PropTypes.func,
    variables: PropTypes.object
  }

  static contextTypes = {
    relayEnvironment: PropTypes.any.isRequired
  }

  render () {
    const {
      container: Container,
      error: ErrorComponent,
      loading: LoadingComponent,
      query,
      render,
      variables,
      ...otherProps
    } = this.props
    const { relayEnvironment } = this.context

    return <QueryRenderer
      query={query}
      environment={relayEnvironment}
      variables={variables}
      render={({ error, props }) => {
        if (render) {
          const combinedProps = { ...otherProps, ...props }
          return render({ error, props: combinedProps })
        } else if (error && ErrorComponent) {
          return <ErrorComponent error={error} {...otherProps} />
        } else if (error) {
          return null
        } else if (props) {
          return <Container {...otherProps} {...props} />
        } else if (LoadingComponent) {
          return <LoadingComponent {...otherProps} />
        } else {
          return null
        }
      }}
    />
  }
}
