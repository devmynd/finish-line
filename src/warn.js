export const warn = (condition, message) => {
  if (condition) {
    console.warn(message)
  }
}
