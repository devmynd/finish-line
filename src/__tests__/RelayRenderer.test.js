/* eslint-env jest */
import React from 'react'
import { mount, shallow } from 'enzyme'
import { RelayRenderer } from '../RelayRenderer'
import { warn } from '../warn'

jest.mock('react-relay', () => {
  const QueryRenderer = (props) => null
  return { QueryRenderer }
})

jest.mock('../warn', () => {
  const warn = jest.fn()
  return { warn }
})

jest.mock('../RelayEnvironmentProvider', () => {
  const RelayEnvironment = ({ children }) => children({
    commitMutation: jest.fn(),
    current: 'relay-environment',
    refresh: jest.fn()
  })

  return {
    RelayEnvironment
  }
})

describe('Basic QueryRenderer props', () => {
  it('passes the appropriate props through to the QueryRenderer', () => {
    const ContainerDummy = () => 'container'
    const ErrorDummy = () => 'error'
    const LoadingDummy = () => 'loading'
    const renderDummy = () => 'render'

    const subject = mount(<RelayRenderer
      container={ContainerDummy}
      query='query'
      variables={{ some: 'variables' }}
      loading={LoadingDummy}
      error={ErrorDummy}
      render={renderDummy}
    />)
    expect(subject).toMatchSnapshot()
  })
})

describe('render passed to QueryRenderer', () => {
  const ContainerDummy = () => 'container'
  const ErrorDummy = () => 'error'
  const LoadingDummy = () => 'loading'
  let subject

  const callRender = (props, renderArg) => {
    subject = mount(<RelayRenderer
      query='query'
      variables={{ some: 'variables' }}
      {...props}
    />)

    return subject.find('QueryRenderer').props().render(renderArg)
  }

  describe('render function given', () => {
    it('calls the given render with the error and props the QueryRenderer render was called with', () => {
      const aProp = 'a prop'
      const props = { some: 'prop' }
      const combinedProps = { ...props, aProp }
      const error = 'error'
      const returnValue = 'something to return'
      const mockRender = jest.fn(() => returnValue)

      const result = callRender({ render: mockRender, aProp }, { error, props })
      expect(result).toEqual(returnValue)
      expect(mockRender).toHaveBeenCalledWith({ error, props: combinedProps })
    })
  })

  describe('no render function given', () => {
    describe('on error', () => {
      it('renders null without a given error component', () => {
        const result = callRender({}, { error: 'an error' })
        expect(result).toEqual(null)
      })

      it('renders the error component when one is given', () => {
        const aProp = 'a prop'
        const error = 'some error'
        const result = callRender({ error: ErrorDummy, aProp }, { error })
        expect(result).toEqual(<ErrorDummy
          error={error}
          aProp={aProp}
          refreshRenderer={subject.instance().refreshRenderer}
        />)
      })
    })

    describe('when given props and no error', () => {
      it('renders the container with the remaining props', () => {
        const aProp = 'a prop'
        const props = { some: 'prop' }
        const result = callRender({ aProp, container: ContainerDummy }, { props })
        expect(result).toEqual(<ContainerDummy aProp={aProp} {...props} />)
      })
    })

    describe('no props and no error', () => {
      it('renders null without a given loading component', () => {
        const result = callRender({}, {})
        expect(result).toEqual(null)
      })

      it('renders the loading comopnent when one is given', () => {
        const aProp = 'a prop'
        const result = callRender({ loading: LoadingDummy, aProp }, {})
        expect(result).toEqual(<LoadingDummy aProp={aProp} />)
      })
    })
  })
})

describe('#refreshRenderer', () => {
  it('increments the childKey that is applied to the direct descendant', () => {
    const subject = shallow(<RelayRenderer query='query' />)
    expect(subject).toHaveState('childKey', 1)
    subject.instance().refreshRenderer()
    expect(subject).toHaveState('childKey', 2)
    subject.instance().refreshRenderer()
    expect(subject).toHaveState('childKey', 3)
  })
})

describe('#componentWillMount', () => {
  describe('given render and any prop that is ignored by render', () => {
    it('logs a warning', () => {
      shallow(<RelayRenderer
        query='query'
        render={() => null}
        container={() => null}
      />)
      expect(warn).toHaveBeenCalledWith(true, 'RelayRenderer was rendered with a `render` prop and `container`, `error`, and/or `loading` props as well. Passing `render` causes those other props to have no affect.')
    })
  })

  describe('given render without props that are ignored by render', () => {
    it('does not log a warning', () => {
      shallow(<RelayRenderer
        query='query'
        render={() => null}
      />)
      expect(warn).toHaveBeenCalledWith(false, expect.anything())
    })
  })

  describe('not given render', () => {
    it('does not log a warning', () => {
      shallow(<RelayRenderer
        query='query'
        container={() => null}
      />)
      expect(warn).toHaveBeenCalledWith(false, expect.anything())
    })
  })
})
