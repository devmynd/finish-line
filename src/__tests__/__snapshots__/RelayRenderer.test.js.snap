// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic QueryRenderer props passes the appropriate props through to the QueryRenderer 1`] = `
<RelayRenderer
  container={[Function]}
  error={[Function]}
  loading={[Function]}
  query="query"
  render={[Function]}
  variables={
    Object {
      "some": "variables",
    }
  }
>
  <QueryRenderer
    environment="relay-environment"
    query="query"
    render={[Function]}
    variables={
      Object {
        "some": "variables",
      }
    }
  />
</RelayRenderer>
`;
